proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2934,6 +2939,105 @@ lemma reply_remove_ccorres:
     (replyRemove replyPtr tcbPtr) (Call reply_remove_'proc)"
 sorry (* FIXME RT: reply_remove_ccorres *)
 
+lemma getBlockingObject_BlockedOnReceive_return:
+  "(getBlockingObject (Structures_H.thread_state.BlockedOnReceive bo bicg ro)) = return bo"
+  unfolding getBlockingObject_def epBlocked_def by simp
+
+lemma getBlockingObject_BlockedOnSend_return:
+  "(getBlockingObject (Structures_H.thread_state.BlockedOnSend bo bib bicg bicgr biic)) = return bo"
+  unfolding getBlockingObject_def epBlocked_def by simp
+
+lemma BlockedOnReceive_replyObject_no_0:
+  assumes "no_0_obj' s"
+  assumes "valid_tcb_state' ts s"
+  assumes "ts = BlockedOnReceive bo bicg ro"
+  shows "ro \<noteq> Some 0"
+  using assms
+  by (auto simp: valid_tcb_state'_def)
+
+lemma ctcb_relation_BlockedOnReceive_blockingObject:
+  assumes "ctcb_relation tcb ctcb"
+  assumes "BlockedOnReceive bo bicg ro = tcbState tcb"
+  shows "blockingObject_CL (thread_state_lift (tcbState_C ctcb)) = bo"
+  using assms
+  by (cases "(tcbState tcb)", simp_all add: ctcb_relation_def cthread_state_relation_def)
+
+lemma ctcb_relation_BlockedOnReceive_replyObject:
+  assumes "ctcb_relation tcb ctcb"
+  assumes "BlockedOnReceive bo bicg ro = tcbState tcb"
+  shows "replyObject_CL (thread_state_lift (tcbState_C ctcb)) = option_to_0 ro"
+  using assms
+  by (cases "(tcbState tcb)", simp_all add: ctcb_relation_def cthread_state_relation_def)
+
+lemma ctcb_relation_BlockedOnSend_blockingObject:
+  assumes "ctcb_relation tcb ctcb"
+  assumes "BlockedOnSend bo bib bicg bicgr biic = tcbState tcb"
+  shows "blockingObject_CL (thread_state_lift (tcbState_C ctcb)) = bo"
+  using assms
+  by (cases "(tcbState tcb)", simp_all add: ctcb_relation_def cthread_state_relation_def)
+
+lemma BlockedOnReceive_ep_at':
+  assumes "valid_objs' s"
+  assumes "st_tcb_at' ((=) (BlockedOnReceive bo bicg ro)) t s"
+  shows "ep_at' bo s"
+  using assms
+  apply -
+  apply (drule (1) tcb_in_valid_state')
+  by (fastforce simp: obj_at'_def valid_tcb_state'_def)
+
+lemma BlockedOnSend_ep_at':
+  assumes "valid_objs' s"
+  assumes "st_tcb_at' ((=) (BlockedOnSend bo bib bicg bicgr biic)) t s"
+  shows "ep_at' bo s"
+  using assms
+  apply -
+  apply (drule (1) tcb_in_valid_state')
+  by (fastforce simp: obj_at'_def valid_tcb_state'_def)
+
+lemma sym_ref_BlockedOnReceive_replyObject_linked:
+  assumes sy: "sym_refs (state_refs_of' s)"
+  assumes vo: "valid_objs' s"
+  assumes st: "st_tcb_at' ((=) (Structures_H.thread_state.BlockedOnReceive bo bicg (Some ro))) thread s"
+  shows "obj_at' (\<lambda>reply. replyTCB reply = Some thread) ro s"
+proof -
+  from vo and st have ra:
+    "reply_at' ro s"
+    by - (drule (1) tcb_in_valid_state', fastforce simp: obj_at'_def valid_tcb_state'_def)
+  from st have ko:
+    "\<exists>tcb. ko_at' tcb thread s \<and> tcbState tcb = BlockedOnReceive bo bicg (Some ro)"
+    by (clarsimp simp: st_tcb_at'_def obj_at'_def)
+  from sy and vo and st and ra and ko show ?thesis
+    apply -
```

Comment:
you don't need `apply -` for tactics like simp/clarsimp/fastforce/auto/etc, it's mostly for direct rule invocations (obvious, I know :/)
you can also shave off the `ko` by using `with` instead of `from`